### PR TITLE
Update headers and a small copy from ts scripts

### PR DESCRIPTION
**Pull request by timlyeee**
* delete Timer Class, keep TimerTargetCallback as the only Timer(should be renamed by the way). Since no-one else extends to Timer.
* replace ISchedulable pointers with System pointers, as targets.update will be called and only cc::system has update function
* rejust some type definitions:
 - Scheduler::_hashForUpdate  map<void *, HashUpdateEntry> -->map<std::string, HashUpdateEntry> 
 - Scheduler::_arrayForTimers vector<Timer*>-->vector<HashTimerEntry*>
* use a copy code as c++ from origin ts script to test and figure out logics, will rewrite if nessecary
* hack release function to release space(should use our own way to release space)

### CODE REVIEW

File: cocos/core/Scheduler.cpp [hunk 1]
```diff
@@ -22,91 +22,694 @@
...
+    typedef A(fnType)(Values...);
+    return (size_t)func.target<fnType*>();
+
+}
+template<class T>
+static void releaseVector(std::vector<T*>& v) {
+    //if it's still in use, then should not consider the capacity
+    for (auto i : v) {
+        delete i;
+        i = nullptr;
```
Comment (minggo) [inline cocos/core/Scheduler.cpp]: 没必要赋值为 null 了

File: cocos/core/Scheduler.cpp [hunk 2]
```diff
@@ -22,91 +22,694 @@
...
+    }v.clear();
+}
+
+template<class F, class S>
+static void releaseMap(std::unordered_map<F, S*>& m) {
+    for (auto i : m) {
+        delete i.second;
+        i.second = nullptr;
```
Comment (minggo) [inline cocos/core/Scheduler.cpp]: 同理

File: cocos/core/Scheduler.h [hunk 3]
```diff
@@ -385,9 +390,9 @@ class Scheduler final : public System {
      * @zh
      * 恢复指定数组中所有对象的定时器。<br/>
      * 这个函数是 pauseAllCallbacks 的逆操作。
-     * @param targetsToResume
+     * @param std::vector<System *> targetsToResume
      */
-    void resumeTargets(ccSchedulerFunc& targetsToResume);
+    void resumeTargets(std::vector<System*>&);
```
Comment (minggo) [inline cocos/core/Scheduler.h]: 加 const ？

**Comment by timlyeee** [comment]
Add director.h

File: cocos/core/Director.h [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+
+    /**
+     * @en
+     * Converts a view coordinate to an WebGL coordinate<br/>
+     * Useful to convert (multi) touches coordinates to the current layout (portrait or landscape)<br/>
+     * Implementation can be found in directorWebGL.
+     * @zh 将触摸点的屏幕坐标转换为 WebGL View 下的坐标。
+     * @deprecated since v2.0
+     */
+    Vec2 convertToGL(Vec2& uiPoint);
```
Comment (minggo) [inline cocos/core/Director.h]: Vec2& uiPoint -> const Vec2& uiPoint
其他的参数也检查一下

File: cocos/core/Director.h [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+    static std::string EVENT_BEGIN_FRAME;
+
+    /**
+     * The event which will be triggered at the frame end.<br/>
+     * 一帧结束之后所触发的事件。
+     * @event Director.EVENT_END_FRAME
+     */
+    static std::string EVENT_END_FRAME;
+
+    static std::shared_ptr<Director> getInstance() {
```
Comment (minggo) [inline cocos/core/Director.h]: 先不要用智能指针，我们还没有决定用智能指针。

File: cocos/core/Director.h [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+    /**
+     * @en
+     * Run a scene. Replaces the running scene with a new one or enter the first scene.<br>
+     * The new scene will be launched immediately.
+     * @zh 运行指定场景。将正在运行的场景替换为（或重入为）新场景。新场景将立即启动。
+     * @param scene - The need run scene.
+     * @param onBeforeLoadScene - The function invoked at the scene before loading.
+     * @param onLaunched - The function invoked at the scene after launch.
+     */
+    void runSceneImmediate(std::variant<Scene, SceneAsset>, std::function<void*()> onBeforeLoadScene, std::function<void*()> onLaunched);
```
Comment (minggo) [inline cocos/core/Director.h]: 确认一下参数的类型。

Comment (minggo) [reply cocos/core/Director.h]: 其他的函数也一样

File: cocos/core/Director.h [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    void resume() { _paused = false; }
+
+    Root* getRoot() { return _root; }
+
+    /**
+     * @en Returns current logic Scene.
+     * @zh 获取当前逻辑场景。
+     */
+    scenegraph::Scene* getScene() { return _scene; }
```
Comment (minggo) [inline cocos/core/Director.h]: 任何 get 类型都要想到 const 
scenegraph::Scene* getScene() -> scenegraph::Scene* getScene() const

File: cocos/core/Director.h [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+     * @param onLoaded 加载回调。
+     */
+    template <typename... A, typename... B>
+    void preloadScene(std::string sceneName, std::function<void*(A...)> onProgress, std::function<void*(B...)> onLoaded);
+
+    /**
+     * @en Resume game logic execution after pause, if the current scene is not paused, nothing will happen.
+     * @zh 恢复暂停场景的游戏逻辑，如果当前场景没有暂停将没任何事情发生。
+     */
+    void resume() { _paused = false; }
```
Comment (minggo) [inline cocos/core/Director.h]: 可以加 inline

File: cocos/core/Director.h [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    /**
+     * @en Run main loop of director
+     * @zh 运行主循环
+     */
+    void tick(float dt);
+	
+    Director();
+    ~Director();
+private:
+    static std::shared_ptr<Director> Instance;
```
Comment (minggo) [inline cocos/core/Director.h]: 参数名不用大写

File: cocos/core/Director.h [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+
+    bool                 _invalid{false};
+    bool                 _paused{false};
+    Root*                _root{nullptr};
+    std::string          _loadingScene;
+    uint32_t             _totalFrames;
+    Scheduler*           _scheduler;
+    std::vector<System*> _systems;
+
+    scenegraph::Scene* _scene;
```
Comment (minggo) [inline cocos/core/Director.h]: 参数默认值，自己多检查一下。

File: cocos/core/Director.h [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+     * @en Returns the scheduler associated with this director.
+     * @zh 获取和 director 相关联的调度器。
+     */
+    inline Scheduler* getScheduler() { return _scheduler; }
+
+    /**
+     * @en Sets the scheduler associated with this director.
+     * @zh 设置和 director 相关联的调度器。
+     */
+    inline void setScheduler(Scheduler* scheduler);
```
Comment (minggo) [inline cocos/core/Director.h]: 如果不在 .h 实现的话，不要加 inline

File: cocos/core/Director.h [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+    static std::string EVENT_END_FRAME;
+
+    static std::shared_ptr<Director> getInstance() {
+        if (!Instance) {
+            Instance = std::make_shared<Director>();
+        }
+        return Instance;
+    }
+
+    scenegraph::ComponentScheduler* _compScheduler;
```
Comment (minggo) [inline cocos/core/Director.h]: 默认值

File: cocos/core/Director.h [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+    /**
+     * The event which will be triggered at the frame end.<br/>
+     * 一帧结束之后所触发的事件。
+     * @event Director.EVENT_END_FRAME
+     */
+    static std::string EVENT_END_FRAME;
+
+    static const Director* getInstance() {
+        if (!instance) {
+            instance = new Director();
```
Comment (minggo) [inline cocos/core/Director.h]: 静态变量的访问尽量用 Director::instance，这样清晰一些。

**Pull request merged**